FWCore/TestModules/plugins/CheckClassInfo.cc

```diff
@@ -0,0 +1,153 @@
+
+  private:
+    void check(edm::ProductDescription const& product) const;
+
+    std::vector<edm::ProductNamePattern> eventProducts_;
+    std::vector<edm::ProductNamePattern> lumiProducts_;
+    std::vector<edm::ProductNamePattern> runProducts_;
+    std::vector<edm::ProductNamePattern> processProducts_;
+  };
+
+  CheckClassInfo::CheckClassInfo(edm::ParameterSet const& config)
+      : eventProducts_(edm::productPatterns(config.getUntrackedParameter<std::vector<std::string>>("eventProducts"))),
+        lumiProducts_(edm::productPatterns(config.getUntrackedParameter<std::vector<std::string>>("lumiProducts"))),
+        runProducts_(edm::productPatterns(config.getUntrackedParameter<std::vector<std::string>>("runProducts"))),
+        processProducts_(
+            edm::productPatterns(config.getUntrackedParameter<std::vector<std::string>>("processProducts"))) {
+    callWhenNewProductsRegistered([this](edm::ProductDescription const& product) {
+      switch (product.branchType()) {
+        case edm::InEvent:
+          for (auto const& label : eventProducts_)
+            if (label.match(product)) {
+              check(product);
+              break;
+            }
+          break;
+
+        case edm::InLumi:
+          for (auto const& label : lumiProducts_)
+            if (label.match(product)) {
+              check(product);
+              break;
+            }
+          break;
+
+        case edm::InRun:
+          for (auto const& label : runProducts_)
+            if (label.match(product)) {
+              check(product);
+              break;
+            }
+          break;
+
+        case edm::InProcess:
+          for (auto const& label : processProducts_)
+            if (label.match(product)) {
+              check(product);
+              break;
+            }
+          break;
+
+        default:
+          throw edm::Exception(edm::errors::LogicError)
+              << "Unexpected branch type " << product.branchType() << "\nPlease contact a Framework developer\n";
+      }
+    });
+  }
+
+  void CheckClassInfo::check(edm::ProductDescription const& product) const {
+    if (product.transient()) {
+      edm::LogVerbatim("CheckClassInfo") << "The product " << product.friendlyClassName() << '_'
+                                         << product.moduleLabel() << '_' << product.productInstanceName() << '_'
+                                         << product.processName() << " is transient, and will not be queried.\n";
+    } else {
+      edm::LogVerbatim log("CheckClassInfo");
+      std::cerr << "product " << product.friendlyClassName() << '_' << product.moduleLabel() << '_'
+                << product.productInstanceName() << '_' << product.processName() << " :\n";
+      std::cerr << "  wrapper type " << product.wrappedType().name() << '\n';
+      TClass* type = product.wrappedType().getClass();
+      std::cerr << "  TClass pointer " << type << '\n';
+      assert(type);
```

Comment:
How about throwing an exception instead of assertion in the case the `type` and `info` pointers are null? The exception message could then contain the information from the `ProductDescription`, which would make it easier to investigate the cases when these checks fail without having to enable the printouts for all types. This could be particularly useful if this module would get used as part of `runTheMatrix`.

Reply:
ok